staging/src/k8s.io/client-go/plugin/pkg/client/auth/azure/azure.go

```diff
@@ -0,0 +1,453 @@
+)
+
+const (
+	azureTimeFormat = "\"2006-01-02 15:04:05.99999\""
+	azureAuthority  = "https://login.microsoftonline.com"
+	uuidFormat      = "[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-4[a-fA-F0-9]{3}-[8|9|aA|bB][a-fA-F0-9]{3}-[a-fA-F0-9]{12}"
+	azureTokenKey   = "azureTokenKey"
+
+	cfgClientID     = "client-id"
+	cfgTenantID     = "tenant-id"
+	cfgAudience     = "audience"
+	cfgAccessToken  = "access-token"
+	cfgRefreshToken = "refresh-token"
+	cfgTokenType    = "token-type"
+	cfgExpiresIn    = "expires-in"
+	cfgExpiresOn    = "expires-on"
+)
+
+func init() {
+	if err := restclient.RegisterAuthProviderPlugin("azure", newAzureAuthProvider); err != nil {
+		glog.Fatalf("Failed to register azure auth plugin: %v", err)
+	}
+}
+
+var cache = newAzureTokenCache()
+
+type azureTokenCache struct {
+	lock  sync.Mutex
+	cache map[string]*azureToken
+}
+
+func newAzureTokenCache() *azureTokenCache {
+	return &azureTokenCache{cache: make(map[string]*azureToken)}
+}
+
+func (c *azureTokenCache) getToken(tokenKey string) *azureToken {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	return c.cache[tokenKey]
+}
+
+func (c *azureTokenCache) setToken(tokenKey string, token *azureToken) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	c.cache[tokenKey] = token
+}
+
+func newAzureAuthProvider(_ string, cfg map[string]string, persister restclient.AuthProviderConfigPersister) (restclient.AuthProvider, error) {
+	var ts tokenSource
+
+	clientID := cfg[cfgClientID]
+	tenantID := cfg[cfgTenantID]
+	audience := cfg[cfgAudience]
+	if clientID != "" && tenantID != "" && audience != "" {
+		ts = newAzureTokenSourceDeviceCode(clientID, tenantID, audience)
+	} else {
+		azFilename, err := azureTokensFile()
+		if err != nil {
+			return nil, err
+		}
+		ts, err = newAzureTokenSourceFromFile(azFilename)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	cacheSource := newAzureTokenSourceFromCache(ts, cache, cfg, persister)
+
+	return &azureAuthProvider{
+		tokenSource: cacheSource,
+	}, nil
+}
+
+type azureAuthProvider struct {
+	tokenSource tokenSource
+}
+
+func (p *azureAuthProvider) Login() error {
+	return errors.New("not yet implemented")
+}
+
+func (p *azureAuthProvider) WrapTransport(rt http.RoundTripper) http.RoundTripper {
+	return &azureRoundTripper{
+		tokenSource:  p.tokenSource,
+		roundTripper: rt,
+	}
+}
+
+type azureRoundTripper struct {
+	tokenSource  tokenSource
+	roundTripper http.RoundTripper
+}
+
+func (r *azureRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	token, err := r.tokenSource.Token()
+	if err != nil {
+		glog.Errorf("Failed to acquire a token: %v", err)
+		return nil, err
+	}
+
+	// clone the request in order to avoid modifying the headers of the original request
+	req2 := new(http.Request)
+	*req2 = *req
+	req2.Header = make(http.Header, len(req.Header))
+	for k, s := range req.Header {
+		req2.Header[k] = append([]string(nil), s...)
+	}
+
+	req2.Header.Set("Authorization", fmt.Sprintf("Bearer %s", token.token.AccessToken))
```

Comment:
I guess I'm still pretty confused as to how this conforms to OIDC, or is secure in its usage. To be clear, as far as I can tell, it seems to be the same pattern that Google uses... but it seems invalid with my understanding of OIDC.

Consider: If the `client_id` is not validated, and any old access token can be used, it means that I can take accessTokens granted to me via *any application* and then use them to talk to your apiserver? That seems... potentially very bad. Or am I misunderstanding? Doesn't this mean my personal application that lets users do OAuth flow to let me create some things in their subscription... can't I now turn around and use those scoped access tokens to get access to some disparate Kubernetes clusters that just happen to be configured for the same tenant without the user's consent?

This is sort of the scenario discussed in the link above that discourages the use of accessTokens for AuthN...